sql: fix small race in distIndexBackfiller

stevendanna opened the pull request:
This fixes a small race condition in
distIndexBackfiller. updateJobDetails calls SetResumeSpansInJob which
mutates the ResumeSpanList in the job details.  Normally, this is only
called from the periodic updater.  However, when the testing knob
AlwasyUpdateIndexBackfillDetails is set, we also update it on every
ProducerMetadata message we get back

Release note: None

stevendanna committed:
sql: fix small race in distIndexBackfiller

This fixes a small race condition in
distIndexBackfiller. updateJobDetails calls SetResumeSpansInJob which
mutates the ResumeSpanList in the job details.  Normally, this is only
called from the periodic updater.  However, when the testing knob
AlwasyUpdateIndexBackfillDetails is set, we also update it on every
ProducerMetadata message we get back

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75857)
<!-- Reviewable:end -->

stevendanna committed:
sql: fix a small race in distIndexBackfill

This is another small data race in distIndexBackfill. This code was
operating on mu.updatedTodoSpans (without holding the lock) instead of
the copy that we created a few lines above (while holding the lock).

Release note: None

stevendanna commented:
@adityamaru Apologies, I pushed another commit for another race that I found when iterating.

For the curious, these are only surfacing now because we happened to write a test that, for $reasons$ was using an incredibly low update interval.

stevendanna commented:
bors r=adityamaru
